config_v2.md

```diff
@@ -0,0 +1,370 @@
+    config['dataset_config'] = dict(crop_size=(30, 30), random_flip=True)
+    config['dataset_size'] = 500
+    
+
+def config_change2(config):
+    config['dataset_config'] = dict(crop_size=(50, 50), random_flip=False)
+    config['dataset_size'] = 700
+    
+    
+def config_change3(config):
+    config_change1(config)
+    config['dataset_size'] = 1_000_000
+
+
+potential_modifs = [config_change1, config_change2, config_change3]
+configuration = Config(dict(batch_size=32, dataset_size=10_000, nb_epochs=50),
+                       potential_modifications=potential_modifs)
+
+
+ex = sacred.Experiment('my_pretty_experiment', config=configuration)
+                       
+def my_main_function(batch_size, dataset_size, nb_epochs, dataset_config):
+    # main experiment here
+    my_dataset = load_dataset(dataset_size, **dataset_config)
+    ...
+
+with ex.start():
+    my_main_function(**configuration)
+```
+
+
+### Example with descriptions of the config values:
+
+
+The `Config` object will remove the `ConfigValues` and keep only the value, for ease of access.
+It will put the descriptions somewhere else, in another attribute of the `Config` object. 
+Let's say `Config.values_descriptions`. 
+
+Notice the way the batch size value is accessed when getting the default dataset size. The ConfigValue has been removed in the constructor.
+
+```python
+import sacred
+from sacred import ConfigValue, Config
+
+
+configuration = Config(dict(
+    batch_size=ConfigValue(32, "The batch size value"),
+    dataset_size=ConfigValue(lambda config: config['batch_size'] * 100, "The dataset size", delayed=True),
+    nb_epochs=ConfigValue(50, "The number of epochs")
+))
+print(configuration['batch_size'])  # will print 32, not a ConfigValue.
+
+ex = sacred.Experiment('my_pretty_experiment', config=configuration)
+                       
+def my_main_function(batch_size, dataset_size, nb_epochs):
+    # main experiment here
+    ...
+
+with ex.start():
+    my_main_function(**configuration)
+```
+
+
+### Example with dynamic population of the potential modifications.
+
+
+There is a hierarchy now. Similar to ingredients.
+
+```python
+import sacred
+from sacred import Config
+
+from somewhere_else import get_mnist, get_cifar
+
+
+def config_mnist1(config):
+    config['dataset_args'] = dict(crop_size=(30, 30), random_flip=True)
+
+
+def config_mnist2(config):
+    config['dataset_args'] = dict(crop_size=(40, 60), random_flip=False)
+
+
+def config_dataset_mnist(config):
+    config['function_get_dataset'] = get_mnist
+    config.add_potential_modification(config_mnist1)
+    config.add_potential_modification(config_mnist2)
+    
+
+
+def config_cifar1(config):
+    config['dataset_args'] = dict(color=True, random_flip=True)
+
+
+def config_cifar2(config):
+    config['dataset_args'] = dict(color=False, random_flip=False)
+
+
+def config_dataset_cifar(config):
+    config['function_get_dataset'] = get_cifar
+    config.add_potential_modification(config_cifar1)
```

Comment:
I'm curious when this is neessary? In this example, couldn't `config_cifar[12]` call `config_dataset_cifar` instead, to make sure that we don't have CIFAR-only arguments set with a non-CIFAR config.

Reply:
The point here is to have a hierarchy and avoid putting 6 functions in the list of `potential_modifications`. It's a toy example and you're right, it's easier to call `config_dataset_cifar` from `config_cifar_[12]`. But then you would have at least 4 functions in the list `potential_modifications`.

Since it's an easy toy example, I'll make a better description in the document to illustrate my point. Thanks for your input.

Reply:
I Added a small description to explain how it reduces the list of potential modifications. Thank you for your input.